samples/bot-tag-mention/csharp/Dialogs/MainDialog.cs

```diff
@@ -0,0 +1,143 @@
+                    Text = "Please Sign In",
+                    Title = "Sign In",
+                    Timeout = 300000, // User has 5 minutes to login (1000 * 60 * 5)
+                    EndOnInvalidMessage = true
+                }));
+
+            AddDialog(new ConfirmPrompt(nameof(ConfirmPrompt)));
+
+            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
+            {
+                PromptStepAsync,
+                MentionTagAsync
+            }));
+
+            // The initial child Dialog to run.
+            InitialDialogId = nameof(WaterfallDialog);
+        }
+
+        private async Task<DialogTurnResult> PromptStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("PromptStepAsync() called.");
+            return await stepContext.BeginDialogAsync(nameof(OAuthPrompt), null, cancellationToken);
+        }
+
+        private async Task<ResourceResponse> TagMentionAdaptivecard(WaterfallStepContext stepContext, CancellationToken cancellationToken, string tagName, string tagId)
+        {
+            var adaptiveCardTemplate = Path.Combine(".", "Resources", "UserMentionCardTemplate.json");
+            var templateJSON = File.ReadAllText(adaptiveCardTemplate);
+            AdaptiveCardTemplate template = new AdaptiveCardTemplate(templateJSON);
+            var memberData = new
+            {
+                tagId = tagId,
+                tagName = tagName
+            };
+            string cardJSON = template.Expand(memberData);
+            var adaptiveCardAttachment = new Attachment
+            {
+                ContentType = "application/vnd.microsoft.card.adaptive",
+                Content = JsonConvert.DeserializeObject(cardJSON),
+            };
+            return await stepContext.Context.SendActivityAsync(MessageFactory.Attachment(adaptiveCardAttachment), cancellationToken);
+        }
+
+        private async Task<DialogTurnResult> MentionTagAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            // Get the token from the previous step. Note that we could also have gotten the
+            // token directly from the prompt itself. There is an example of this in the next method.
+            var tokenResponse = (TokenResponse)stepContext.Result;
+            bool tagExists = false;
+            if (tokenResponse?.Token != null)
+            {
+                try
+                {
+                    stepContext.Context.Activity.RemoveRecipientMention();
+                    if (stepContext.Context.Activity.Text.Trim().ToLower().Contains("<at>"))
+                    {
+                        var tagName = stepContext.Context.Activity.Text.Replace("<at>", string.Empty).Replace("</at>", string.Empty).Trim();
+                        var tagID = stepContext.Context.Activity.Entities[1].Properties.First.First.First.First.ToString();
```

Comment:
A more straightforward way to get id is:
tagId = stepContext.Context.Activity.Entities[1].Properties["mentioned"]["id"]ToString();

Reply:
Thanks for mentioning it, I have fixed it now.